STM32Cube/Tasks/vn_handler.c

```diff
@@ -0,0 +1,143 @@
+ *      Author: joedo
+ */
+
+
+#include "vn_handler.h"
+#include "vn/protocol/upack.h"
+#include "FreeRTOS.h"
+#include "semphr.h"
+#include "fake_logging.h"
+#include "stm32h7xx_hal.h"
+
+
+//RateDivisor sets the output rate as a function of the sensor ImuRate (800 Hz for the VN-200)
+#define CONFIG_MSG_LENGTH 100
+#define VN_200_IMURATE 800 //Hz
+
+//Binary Output #1
+#define BINARY1_OUT_RATE 10 //Hz
+#define RATE1_DIVISOR VN_200_IMURATE / BINARY1_OUT_RATE
+
+//See "vnenum.h" and the ICD for parameter definitions
+/*
+#define COMMONGROUP_PARAMS COMMONGROUP_QUATERNION | COMMONGROUP_POSITION | COMMONGROUP_VELOCITY | COMMONGROUP_ACCEL
+#define TIMEGROUP_PARAMS TIMEGROUP_TIMEUTC | TIMEGROUP_TIMEUTC
+#define IMUGROUP_PARAMS IMUGROUP_IMUSTATUS | IMUGROUP_SENSSAT | IMUGROUP_UNCOMPMAG | IMUGROUP_UNCOMPACCEL | IMUGROUP_UNCOMPGYRO | IMUGROUP_TEMP | IMUGROUP_PRES
+#define GPSGROUP_PARAMS GPSGROUP_UTC | GPSGROUP_NUMSATS | GPSGROUP_FIX | GPSGROUP_POSLLA | GPSGROUP_VELECEF
+#define ATTITUDEGROUP_PARAMS ATTITUDEGROUP_NONE
+#define INSGROUP_PARAMS INSGROUP_NONE
+*/
+
+//Minimize binary output contents for initial testing
+#define COMMONGROUP_PARAMS COMMONGROUP_NONE
+#define TIMEGROUP_PARAMS TIMEGROUP_TIMEUTC
+#define IMUGROUP_PARAMS IMUGROUP_NONE
+#define GPSGROUP_PARAMS GPSGROUP_NONE
+#define ATTITUDEGROUP_PARAMS ATTITUDEGROUP_NONE
+#define INSGROUP_PARAMS INSGROUP_NONE
+
+
+#define MAX_BINARY_OUTPUT_LENGTH 200
+#define DMA_RX_TIMEOUT 300
+uint8_t USART1_Rx_Buffer[MAX_BINARY_OUTPUT_LENGTH];
+SemaphoreHandle_t USART1_DMA_Sempahore;
+
+void USART1_DMA_Rx_Complete_Callback(UART_HandleTypeDef *huart)
+{
+	BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+	xSemaphoreGiveFromISR(USART1_DMA_Sempahore, &xHigherPriorityTaskWoken);
+	portYIELD_FROM_ISR( xHigherPriorityTaskWoken );
+}
+
+int vnIMUSetup(){
+	//Setup DMA on Rx
+	HAL_UART_RegisterCallback(&huart1, HAL_UART_RX_COMPLETE_CB_ID, USART1_DMA_Rx_Complete_Callback);
+	//TODO: Read the WHOAMI registers or equivalent to verify IMU is connected, if not throw error
+		//Register ID 0x01: Offset 0 Model string[24] – Product model number, maximum length 24 characters.
+
+	//configure binary output for raw IMU data and full state estimate
+	uint8_t buffer[CONFIG_MSG_LENGTH];
+	size_t returnedLength;
+	VnError err = VnUartPacket_genWriteBinaryOutput1(buffer, CONFIG_MSG_LENGTH, VNERRORDETECTIONMODE_CHECKSUM, &returnedLength, ASYNCMODE_BOTH, RATE1_DIVISOR,COMMONGROUP_PARAMS,TIMEGROUP_PARAMS,IMUGROUP_PARAMS,GPSGROUP_PARAMS,ATTITUDEGROUP_PARAMS,INSGROUP_PARAMS, GPSGROUP_NONE);
+	if (err == E_NONE)
+	{
+		return HAL_UART_Transmit(&huart1, buffer, returnedLength, 10); //Send the command to configure IMU output format
+		//TODO: Log error to logQueue
+	}
+	else
+	{
+		//TODO: Log error to logQueue
+	}
+	return -1;
+}
+
+
+void vnIMUHandler(void *argument)
+{
+	USART1_DMA_Sempahore = xSemaphoreCreateBinary();
+	for(;;)
+	{
+		xSemaphoreTake(USART1_DMA_Sempahore, 0); //Take semaphore with no timeout to ensure state
```

Comment:
What does this mean "to ensure state"? Why are we trying to take empty semaphore

Reply:
They do it in an example in pg 259 of the book but that example probably isn't relevant. Happened to be on my mind at the time when writing i guess. I will change it

Reply:
not entirely sure what your intention is here, but this appears to still be polling w 0 timeout before making the request